### PR TITLE
BL-72 -fe-as-a-user-i-can-filter-asylum-case-decisions-by-one-or-more-columns-from-the-dataset

### PR DESCRIPTION
**Pull request by VWashingtonCoder**
## Description

BL-72 FE
As A User I Can Filter Asylum Case Decisions By One Or More Columns From The Dataset

Search/Filter functionality for columns/categories has been added within reducers, actions, and a new search bar file. Updated table to show results of search/filter. Adding this feature has made it easier for users to find specific information based on their needs.

#### Video Link

[Loom Video](https://www.loom.com/share/a90c4741fc1147068cc61b79b2f7b548)

#### Jira Link

[BL-72] (https://bloomtechlabs.atlassian.net/jira/software/c/projects/BL/boards/8?modal=detail&selectedIssue=BL-72)

## Type of change

___Please delete options that are not relevant.___

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have removed unnecessary comments/console logs from my code
- [X] My changes generate no new warnings
- [X] I have checked my code and corrected any misspellings
- [X] No duplicate code left within changed files
- [X] Size of pull request kept to a minimum
- [X] Pull request description clearly describes changes made & motivations for said changes

### CODE REVIEW

**Review by EstrellaDionis — DISMISSED** [review]
Very well done and I think will make for great user experience and the code is very clean and easy to read.

**Review by Jason7jl — COMMENTED** [review]
Great job

**Review by attaris978 — COMMENTED** [review]
Elegant, well structured, and readable.
A note on best practices: when using a switch statement, you can enclose the "do-that" part of each "if-this : do that" in squiggly brackets. While unnecessary in this particular instance, it provides a couple of benefits, foremost of which is that it sets each "do-that" inside its own block with its own block scope, which prevents accidental modification of variables. If we don't create these block scopes, then the scope of each case is going to be the entire switch statement, which can have unintended consequences.
So instead of:
case "some condition" : return "wow, what a condition!"
You could simply write:
case "some condition" : { return "wow, what a condition!" }

File: src/state/reducers/dataReducer.js [hunk 1]
```diff
@@ -1,16 +1,67 @@
...
     case GET_DATA:
       return action.payload;
     default:
       return state;
   }
-};
+}
+
+function filteredReducer(state = initialFilteredStates, action) {
+  switch (action.type) {
```
Comment (attaris978) [inline src/state/reducers/dataReducer.js:42]: Consider wrapping the resulting operation from each switch case in squiggly brackets in order to create block scope and prevent unintended modification of variable values which might result from too expansive a scope(presently, the scope for each statement is the entire switch statement)

**Comment by billydean** [comment]
Quick note and then a comment.
Note: Remember to assign the pull request to the owner (and co-owner, if applicable), as well as to anyone who contributed (i.e. on subtasks, if any). 

Comment: This didn't occur to me until after our Product Review today, but we might need to rethink how filtering-by-category works in this table.  The user should be able to filter using a search query (as you've already provided for), but they should also be able to filter by certain broad, pre-set categories -- the example from our product guide and design documents is "by region." I don't think we've gotten a list of those regional (or other) categories from our stakeholder yet, though.

Long story short: I think what you've done for category filters is fine for now, structurally, but we'll have to tweak it to account for the stakeholder's desired categories.

**Review by billydean — CHANGES_REQUESTED** [review]

**Comment by billydean** [comment]
> Quick note and then a comment. Note: Remember to assign the pull request to the owner (and co-owner, if applicable), as well as to anyone who contributed (i.e. on subtasks, if any).
> 
> Comment: This didn't occur to me until after our Product Review today, but we might need to rethink how filtering-by-category works in this table. The user should be able to filter using a search query (as you've already provided for), but they should also be able to filter by certain broad, pre-set categories -- the example from our product guide and design documents is "by region." I don't think we've gotten a list of those regional (or other) categories from our stakeholder yet, though.
> 
> Long story short: I think what you've done for category filters is fine for now, structurally, but we'll have to tweak it to account for the stakeholder's desired categories.

Oops -- Scratch that! A lot of the above is more relevant for BL-71 and BL-73. What you have here is totally in line with the ticket's user story. I guess I ought to redirect this comment to one of our managers (or bring it up at our next product review).

File: src/components/pages/Table/SearchBar.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+          onChange={onChange}
+        ></input>
+      </label>
+      <select
+        onChange={onChange}
+        value={form.category}
+        name="category"
+        id="category"
+      >
+        <option value="">Pick A Category</option>
```
[the remark this answers is not in the document]
Comment (VWashingtonCoder) [reply src/components/pages/Table/SearchBar.js:68]: Can I ask what the importance of more than one category at one search point is over the filtering process now? For instance, if searching for a specific asylum office and a specific ethnicity/race, what would be the point of entering more than one search category at a time? 

Unless there are also plans for a second pop-up search bar to take in that second search value as well. Otherwise, I believe it would just be too complicated for regular asylum users. 

If this is to account for things like the hidden parameters, I believe those could be added at a later date when we have our full official data would be.

File: src/state/reducers/dataReducer.js (hunk 1 above)
Comment (VWashingtonCoder) [reply src/state/reducers/dataReducer.js:44]: I went ahead and updated the dataReducer switch cases based on our recommendation.

**Comment by VWashingtonCoder** [comment]
Decided work on this branch fits description of BL-71 instead. Will work on advanced search/filter for BL-72.
